### DISCUSSION TITLE
Stale branches

### DISCUSSION THREAD

**Discussion by FloEdelmann** [streetcomplete/StreetComplete#3662] [Repository Maintenance] [answered]
There are some [stale branches](https://github.com/streetcomplete/StreetComplete/branches/stale) that seem to be unneeded. If possible, I'd like to clean up there to …

* … make sure that no relevant code is lost in stale branches, and …
* … remove mental burden from both new and already-active developers.

I know it's not super high priority though, so just say so if it should stay as it is.

### Feature branches

![Bildschirmfoto am 2022-01-20 um 19 15 49](https://user-images.githubusercontent.com/202916/150398087-faa76f0e-d21b-4ad4-8b79-c05512116392.png)

@smichel17 [`satellite-down`](https://github.com/streetcomplete/StreetComplete/branches/all?query=satellite-down) is still needed for https://github.com/streetcomplete/StreetComplete/pull/2794#issuecomment-940177225, right?

@westnordost What about [`weak-ref-to-tangram`](https://github.com/streetcomplete/StreetComplete/branches/all?query=weak-ref-to-tangram) and [`fetch-by-bbox`](https://github.com/streetcomplete/StreetComplete/branches/all?query=fetch-by-bbox)? They seem to be related to the refactoring in [v32.0](https://github.com/streetcomplete/StreetComplete/releases/tag/v32.0). I tried merging them (locally) into the master branch, but there were too many conflicts. I also think that the *code changes* actually were merged into the master branch, but the *commits* on these branches were not. Maybe some rebasing / merge conflict was happening back then.

### Release branches

![Bildschirmfoto am 2022-01-20 um 19 32 19](https://user-images.githubusercontent.com/202916/150400465-18b8371b-3f4c-4856-a18c-60028e8cae46.png)

There are release branches for many (but not all!) releases.

Except for the very first ([`0`](https://github.com/streetcomplete/StreetComplete/commits/0)) and the very last branch ([`40`](https://github.com/streetcomplete/StreetComplete/commits/40)), the latest commit on each branch is tagged:

* `0`: https://github.com/streetcomplete/StreetComplete/releases/tag/v0.13 + [2 more commits](https://github.com/streetcomplete/StreetComplete/compare/v0.13...0)
* `1`: https://github.com/streetcomplete/StreetComplete/releases/tag/v1.4
* `2`: https://github.com/streetcomplete/StreetComplete/releases/tag/v2.4
* `4`: https://github.com/streetcomplete/StreetComplete/releases/tag/v4.1
* `5`: https://github.com/streetcomplete/StreetComplete/releases/tag/v5.2
* `6`: https://github.com/streetcomplete/StreetComplete/releases/tag/v6.1
* `8`: https://github.com/streetcomplete/StreetComplete/releases/tag/v8.4
* `20`: https://github.com/streetcomplete/StreetComplete/releases/tag/v20.2
* `21`: https://github.com/streetcomplete/StreetComplete/releases/tag/v21.2
* `24`: https://github.com/streetcomplete/StreetComplete/releases/tag/24.2 and https://github.com/streetcomplete/StreetComplete/releases/tag/v24.2
* `27`: https://github.com/streetcomplete/StreetComplete/releases/tag/v27.2
* `32`: https://github.com/streetcomplete/StreetComplete/releases/tag/v32.2
* `34`: https://github.com/streetcomplete/StreetComplete/releases/tag/v34.2
* `35`: https://github.com/streetcomplete/StreetComplete/releases/tag/v35.2
* `36`: https://github.com/streetcomplete/StreetComplete/releases/tag/v36.1
* `37`: https://github.com/streetcomplete/StreetComplete/releases/tag/v37.2
* `38`: https://github.com/streetcomplete/StreetComplete/releases/tag/v38.2

I tried to merge branches `38`, `37` and `36` into `master` (locally), and there were some easy-to-solve merge conflicts. However, the resolution was to always accept the changes from `master`.

@westnordost So I guess we can safely delete these branches? The commits are not lost anyway since they're tagged. And we can probably also delete branch `0` because of its age?

**Comment by westnordost** [comment, marked as answer]
> satellite-down 

yes

> weak-ref-to-tangram

I figure this can be deleted. It was spawned by this issue: https://github.com/tangrams/tangram-es/issues/2265 , or #2879: 
Updating map data from a background thread may result in a crash when at that point, the view has already been destroyed. 
The changes on that branch would solve that theoretical cause for a crash at the cost of readability/higher complexity but either I solved it in another way (i.e. map data is guaranteed to updated on a background thread) or this is only a theoretical concern (don't remember). In any case, I don't see a significant amount of tangram-releated crashes on google play console. (By the way, would you like to have access to it? You can see the crash reports, you can see statistics etc. this way, Mateusz has access too)

> fetch-by-bbox 

was an attempted performance optimization / experiment. It is related to #3609 which is why I would like to keep it around a little longer until a conclusion is reached in that ticket.

---

Also, the numbered branches are the branches for the releases. They are not feature branches and are not meant to be merged into master. They exist so that it is easy to cherry-pick certain fixes from master when master has moved on, so we can make bugfix-updates of old releases.
This mechanism is only used only really for the last version, e.g. if v99 is currently in beta, there'll be a `98` branch on which we can put bugfixes to maybe release a v98.3.
But it can be used if one version for example is decleared a long term support version that should receive bugfix updates longer. For example, the last release before upping the `minSdkVersion` to 21 could have been such a case. Or, in a coming release, I'll ask an additional permission (= the camera permission) of the user, which may let many users get stranded on an older major release (for a while) because they don't want to grant that permission / Google Play doesn't automatically update an app if it asks new permissions. So that would be a case for an LTS too.

I figure we **could** change how we organize this. But what is your suggestion exactly and how would it make it better?

**Reply by FloEdelmann** [reply]
Thanks for the insights. Well I guess it's okay to keep all three feature branches for now.

> Also, the numbered branches are the branches for the releases. They are not feature branches and are not meant to be merged into master. They exist so that it is easy to cherry-pick certain fixes from master when master has moved on, so we can make bugfix-updates of old releases.

Yes, I know. But I wasn't sure if maybe some bug fix commits went into the release branch only.

I suggest to delete release branches older than the current stable version (i.e. now branches `0`, `1`, … `37`, `38`). If we decide in the future to make one of these versions an LTS release (which seems unlikely), we can again create a branch from the tagged commit.

If in the future such a "breaking change" occurs, we can decide to not delete that branch when the next stable version is released.

**Reply by FloEdelmann** [reply]
> I figure we could change how we organize this. But what is your suggestion exactly and how would it make it better?

It sounds like we're using exactly the [GitLab Flow](https://www.gitkraken.com/learn/git/best-practices/git-branch-strategy) branch strategy. I don't think we should change it, it seems to work well for this project.

But maybe we should also document that in the new release checklist you created a few days ago?
https://github.com/streetcomplete/StreetComplete/blob/master/res/documentation/creating%20new%20release.md

**Reply by westnordost** [reply]
I create a release branch only if i need one, i.e. master already moved on but a bifocal update on the last major update should be done. So,  that wouldn't be part of releasing a new version,  just the tagging of the commit is.

**Reply by westnordost** [reply]
So, does this gitlab flow guideline suggest to delete old stable branches?

**Reply by FloEdelmann** [reply]
> that wouldn't be part of releasing a new version

True. I thought about a section like "After the release" with a content like: "If need for minor release and other stuff already on master, then create release branch based off of the tagged commit and cherry-pick the fixes there."

> does this gitlab flow guideline suggest to delete old stable branches?

It doesn't. Surprisingly, it's quite hard to find tips and suggestions about if and when to delete old release branches in any branch strategy. But I did find a StackExchange answer I do agree with: https://softwareengineering.stackexchange.com/a/361817

**Reply by westnordost** [reply]
You are invited to add this to the document if you think it makes sense.

Also, I guess I have no case against deleting old release branches, so if you find it is more clear this way, go ahead and delete them

**Reply by FloEdelmann** [reply]
I added documentation in https://github.com/streetcomplete/StreetComplete/commit/92e4aef9c40102585d7cf8465a3db78f71cd1ce9 and deleted the stale release branches.

**Comment by smichel17** [comment]
Maybe even the in progress (but not actively being worked on) branches would be better as *annotated* tags, since then they can include an explanation of what they're for.

**Reply by FloEdelmann** [reply]
Hmmm, not sure about that. It's difficult to get rid of unwanted tags again, see e.g. https://github.com/refined-github/refined-github/pull/3421#issuecomment-690490844.

I think it would be better to open draft PRs that reference the issue. Then it's also easy to delete the branch when the draft PR is not needed anymore.

**Comment by FloEdelmann** [comment]
> By the way, would you like to have access to it? You can see the crash reports, you can see statistics etc. this way, Mateusz has access too

@westnordost I think I don't need access there. It sounds like another time killer for me (diving into statistics and so on :sweat_smile:) without much benefit. If you make the crash reports public in GitHub issues, all contributors including me can try to fix them. However, I can't help with most bugs anyway.

**Reply by westnordost** [reply]
Ok

**Comment by FloEdelmann** [comment]
@westnordost Are the `create-vertices`, `create-vertices-experiment1` and `create-vertices-experiment2` branches still needed? Probably not, since StreetComplete is able to create new nodes since [v48.0](https://github.com/streetcomplete/StreetComplete/releases/tag/v48.0), right? Can we delete [these stale branches](https://github.com/streetcomplete/StreetComplete/branches/all?query=create-vertices) then?

![grafik](https://user-images.githubusercontent.com/202916/212535941-d0942de2-297f-4817-85be-da62b60dab66.png)

**Comment by westnordost** [comment]
Yes, they are needed

On January 15, 2023 11:40:45 AM GMT+01:00, Flo Edelmann ***@***.***> wrote:
>@westnordost Are the `create-vertices`, `create-vertices-experiment1` and `create-vertices-experiment2` branches still needed? Probably not, since StreetComplete is able to create new nodes since [v48.0](https://github.com/streetcomplete/StreetComplete/releases/tag/v48.0), right? Can we delete these stale branches then?
>
>https://github.com/streetcomplete/StreetComplete/branches/all?query=create-vertices
>
>![grafik](https://user-images.githubusercontent.com/202916/212535941-d0942de2-297f-4817-85be-da62b60dab66.png)
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/discussions/3662#discussioncomment-4689036
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.
